### ISSUE TITLE
categories à la folds

### ISSUE THREAD

**Issue by benediktahrens** [UniMath/UniMath#309]
I once formalized categories a la folds, following notes by Mike Shulman.
The code is in a github repository [1], together with some latex notes.
It compiles with current UniMath.

I am not working on this right now, but i would like the code to remain compatible with UniMath.
Is it ok if I store it in a subdir of CT? That would make that task much less difficult.
It is about 1000 lines of what I consider low-maintenance code.

[1] https://github.com/benediktahrens/folds/tree/master/coq

EDIT: compile time is 7.5 seconds using one core on my rather old laptop.

**Comment by andrejbauer** [comment]
In HoTT we have a `contrib` directory where such stuff goes.

**Comment by DanGrayson** [comment]
It's okay with me.  (It would be a "package".)

**Comment by benediktahrens** [comment]
Ok, see PR #310 .

**Comment by vladimirias** [comment]
Sounds like a good idea to me. Are there any difficulties for you Benedikt to manage it if it is in the contrib subdirectory?

**Comment by benediktahrens** [comment]
On 03/05/2016 06:03 PM, Vladimir Voevodsky wrote:
> Sounds like a good idea to me. Are there any difficulties for you
> Benedikt to manage it if it is in the contrib subdirectory?


At the moment, there is no "contrib" subdirectory in the UniMath
repository. So the answer depends on where this to-be-created contrib
subdirectory will be located.


If "contrib" will be inside the subdirectory UniMath/ of the UniMath
repository, then the current Makefile should work just as well, I think.


If, on the other hand, the "contrib" directory should be outside the
UniMath/ subdirectory, then the current Makefile would need to be
amended to an extent beyond my understanding of Make.
Alternatively, it might be easier to have a separate Makefile for the
Folds package, and have the main Makefile call make on that separate
Makefile.

Also, other benefits would be lost, e.g., automatic loading of Agda
input mode, which only works for Coq files in the the subdirectory UniMath/.

**Comment by benediktahrens** [comment]
I have to amend my answer:

in the second case, the CategoryTheory package (and the Foundations
package) would need to be *installed* in order to make it available to
the Folds package.

For now, we do not have the possibility to install individual packages,
only installation of the whole of UniMath.

However, compiling the whole of UniMath takes too long on Travis, so
that we cannot install UniMath on the Travis machines.

Hence we cannot run the compilation test for the Folds package on Travis
- and this would be the main purpose of putting Folds into the UniMath
repository.

**Comment by DanGrayson** [comment]
We should find a way to pay travis to let us use more cpu time or set up automated testing
ourselves on a machine of our own.  We can't let a travis limitation restrain the
growth of UniMath.

**Comment by DanGrayson** [comment]
I think of UniMath as being the contrib directory, and the individual packages, such Ktheory, as being the contributions.

**Comment by benediktahrens** [comment]
On 03/05/2016 07:48 PM, Daniel R. Grayson wrote:
> We should find a way to pay travis to let us use more cpu time or set up automated testing
> ourselves on a machine of our own.  We can't let a travis limitation restrain the
> growth of UniMath.

I am trying to make a Contrib directory work by extending the loadpath
using the -Q option, but I get a weird error message on Travis; see
around line 5271 on [1].

It works on my machine, I dunno why Travis complains...

[1] https://travis-ci.org/benediktahrens/UniMath/jobs/113976097

**Comment by DanGrayson** [comment]
I don't see any point in having a contrib directory outside the UniMath directory.  Why bother with it?

**Comment by DanGrayson** [comment]
I see in your attempt to make a contrib directory you have added a makefile.  That is not good, because it will make keeping track of dependencies harder.  Stick to one makefile.

**Comment by DanGrayson** [comment]
Perhaps that _Coqproject file is screwing things up.

**Comment by benediktahrens** [comment]
Well, that _CoqProject file seems to be the recommended way of doing things.
It is read by PG, so that PG passes the correct arguments to coqtop. And
using coq_makefile on it is standard, too.

I don't know how to make the Contrib directory outside the UniMath
directory work...

**Comment by benediktahrens** [comment]
On 03/05/2016 07:58 PM, Daniel R. Grayson wrote:
> Perhaps that _Coqproject file is screwing things up.

I don't think so.
What happens is that on the Travis machine, the UniMath modules are not
found in the loadpath, even though the UniMath directory was bound to
"UniMath" with option
-Q ../UniMath UniMath
The very same thing works fine on my machine.


I am giving up my efforts on making a Contrib directory work, and reopen
my original PR.

**Comment by vladimirias** [comment]
Does Folds follow the UniMath discipline? I mean first of all no use of the “inductive” or any of its analogs outside of the prelim.v .

**Comment by vladimirias** [comment]
I am reasonably sure we should be able to use the School of Math machine(s) for this purpose provided that someone (Dan and I? I, to add authority to our requests so that they are processed faster) works with Kevin and Theresa to set it up.

**Comment by benediktahrens** [comment]
On 03/06/2016 06:33 AM, Vladimir Voevodsky wrote:
> Does Folds follow the UniMath discipline? I mean first of all no use
> of the “inductive” or any of its analogs outside of the prelim.v .

Yes. All the code I submit for inclusion in UniMath does, to the best of
my knowledge.

**Comment by benediktahrens** [comment]
I think that my (and in consequence Dan's) comments are misleading.


The current setup (Github + Travis CI) works very well for UniMath,
provided that we stick to the current way of doing things, by which I
mean: everything is a package in the UniMath directory.
If we continue to work that way, I think that Github+Travis will
continue to work for us for many years.


The same setup (Github + Travis CI) did use to not work so well for
projects (i.e., for repositories) that are based on UniMath, i.e., that
require UniMath installed.

(The reason for that is that the install target of the UniMath Makefile
is monolithic, so that we can compile, but not install, individual
packages. For installation, all of UniMath needs to be compiled.)

However, Travis machines seem to be fast enough at the moment to be able
to compile and install the whole of UniMath, and, in addition, to
compile the project by Lumsdaine, Voevodsky and myself, in 48 minutes
(i.e., in the timeframe of 50 minutes allowed).
This is quite tight, and problems might arise in the near future.


To summarize, the problem arises not with the UniMath repository (if we
stick to having only packages, no Contrib stuff).
It arises instead with projects that want UniMath installed.


Moving UniMath on another infrastructure will not help at all to solve
that problem.


A relatively light-weight solution would be to provide a second,
alternative, Makefile that allows installation of individual packages.
Other projects could then set up their Travis configuration such that
make is called on that alternative Makefile.

**Comment by DanGrayson** [comment]
If the problem arises from github projects using both travis and UniMath, then perhaps the
best solution is to provide a compiled UniMath that can be installed by a travis script, as well
as a compiled coq.  We would make a debian repository with *.deb files, and the travis script
could get them from there.  Even better would be to have UniMath be distributed as part of
Ubuntu, which, I think, is what travis' virtual machines use.

**Comment by benediktahrens** [comment]
I fail to see how any of the solutions you propose are maintainable, let
alone easier to maintain than a second Makefile that provides
package-wise install targets.

**Comment by DanGrayson** [comment]
Yes, it's hard to maintain, but the advantage is clear in terms of saving time during the travis build.

Shall I look into making package-wise install targets, or would you like to do that?  They should
definitely be in the main makefile, not in a second one, if at all possible.  You saw what confusion
results from a second makefile when testing that one pull request.

**Comment by benediktahrens** [comment]
Done in PR #310 , closing.
